executor/join.go

```diff
@@ -889,12 +913,37 @@ func (e *NestedLoopApplyExec) Next(ctx context.Context, req *chunk.Chunk) (err e
 			e.hasMatch = false
 			e.hasNull = false
 
-			for _, col := range e.outerSchema {
-				*col.Data = e.outerRow.GetDatum(col.Index, col.RetType)
-			}
-			err = e.fetchAllInners(ctx)
-			if err != nil {
-				return err
+			if e.canUseCache {
+				var keyByte []byte
+				for _, col := range e.outerSchema {
+					*col.Data = e.outerRow.GetDatum(col.Index, col.RetType)
+					keyByte, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, keyByte, *col.Data)
+					if err != nil {
+						return err
+					}
+				}
+				e.totalNumber++
+				key := string(keyByte)
+				value := e.cache.Get(key)
+				if value != nil {
+					e.innerList = value.Data
+					e.cacheHitNumber++
+				} else {
+					err = e.fetchAllInners(ctx)
+					if err != nil {
+						return err
+					}
+					innerList := e.innerList.Copy()
```

Comment:
We should attach the new cloned list's `memTracker` to the apply's `memTracker`.